src/Microsoft.AspNet.OData/Query/PlainTextODataQueryOptionsParser.cs

```diff
@@ -0,0 +1,70 @@
+﻿// Copyright (c) Microsoft Corporation.  All rights reserved.
+// Licensed under the MIT License.  See License.txt in the project root for license information.
+
+using System.IO;
+using System.Diagnostics.CodeAnalysis;
+using Microsoft.OData;
+using Microsoft.AspNet.OData.Common;
+
+namespace Microsoft.AspNet.OData.Query
+{
+    [SuppressMessage("Microsoft.Naming", "CA1702:CompoundWordsShouldBeCasedCorrectly", Justification="Spelling correct in this context")]
+    public partial class PlainTextODataQueryOptionsParser
+    {
+        /// <inheritdoc/>
+        public string Parse(Stream requestStream)
+        {
+            string queryString = string.Empty;
+
+            MemoryStream memoryStream = new MemoryStream();
+            TextReader reader;
+
+            try
+            {
+                // Reset request stream position if possible
+                if (requestStream.CanSeek)
+                {
+                    requestStream.Position = 0;
+                }
+
+                requestStream.CopyTo(memoryStream);
+                memoryStream.Position = 0;
+                reader = new StreamReader(memoryStream);
+            }
+            catch
+            {
+                memoryStream.Dispose();
+
+                throw new ODataException(SRResources.CannotParseQueryOptionsPayload);
+            }
+
+            using (reader)
+            {
+                try
+                {
+                    // Based on OData OASIS Standard, the request body is expected to contain the query options part of OData URL 
+                    // and MUST use the same percent-encoding as in URLs (especially: no spaces, tabs, or line breaks allowed) 
+                    // and MUST follow the expected syntax rules
+
+                    // For consideration: support multiline
+                    // - Would it be a single query option per line?
+                    // - Would the parser be responsible for adding the & separator?
+                    // - Would the parser try to detect the & separator and add where necessary?
+                    string result = reader.ReadToEnd();
```

Comment:
Why isn't reading the stream async?

Reply:
In Microsoft.AspNet.OData project, asynchronous processing is not 

> Also, could you show an example of how a user would register a new custom query options parser?

@habbes Here's how:
```c#
IList<IODataQueryOptionsParser> queryOptionsParsers = ODataQueryOptionsParserFactory.Create();
queryOptionsParsers.Insert(0, new CustomODataQueryOptionsParser());

app.UseMvc(routeBuilder =>
{                
	// ...
	routeBuilder.MapODataServiceRoute(
		routeName: "odata",
		routePrefix: "odata",
		configureAction: containerBuilder => containerBuilder
			// ...
			.AddService(Microsoft.OData.ServiceLifetime.Singleton, typeof(IEnumerable<IODataQueryOptionsParser>), _ => queryOptionsParsers), 
			// ...
	);
	// ...
});
```

Reply:
@habbes Because async processing wasn't/isn't a design consideration for Microsoft.AspNet.OData project

Reply:
@habbes I eventually with async processing to address an issue where a particular test was failing intermittently